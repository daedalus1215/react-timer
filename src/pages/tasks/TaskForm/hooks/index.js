export { default as useFetchTags } from './useFetchTags';
export { default as useTagContext } from './useTagContext';
export { default as useTagTransformer } from './useTagTransformer';
export { default as useTimeContext } from './useTimeContext';
export { default as useSubmit } from './useSubmit/useSubmit';
export { default as useFormReducer } from './useFormReducer';
