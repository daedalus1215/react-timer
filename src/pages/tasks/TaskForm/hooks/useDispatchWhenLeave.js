import { useEffect } from 'react';

const useDispatchWhenLeave = dispatch => {
    return useEffect(() => {
        window.addEventListener('beforeunload', dispatch);
        return () => {
            window.removeEventListener('beforeunload', dispatch);
        };
    }, [dispatch]);
};

export default useDispatchWhenLeave;
// window.addEventListener("beforeunload", (ev) => {
//     ev.preventDefault();
//     return ev.returnValue = 'Are you sure you want to close?';
// });