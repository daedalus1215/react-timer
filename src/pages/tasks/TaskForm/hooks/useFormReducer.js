import { useReducer } from 'react';

const initialState = {
  id: -1,
  time: 0,
  dateTimes: []
};

const formReducer = (state, newState) => {
  return {
    ...state,
    ...newState,
  };
};

const useFormReducer = () => {
  const [state, dispatch] = useReducer(formReducer, initialState);

  const onDateTimes = e => {
    dispatch({ dateTimes: e })
  }

  return [
    state,
    onDateTimes,
  ];
};

export default useFormReducer;
