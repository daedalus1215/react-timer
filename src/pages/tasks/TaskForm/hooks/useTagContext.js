import { useContext } from 'react';
import { TagContext, TagActionContext } from '../contexts/TagContext';

const useTags = () => useContext(TagContext);
const useSetTags = () => useContext(TagActionContext);

const useTagContext = () => {
    const allTags = useTags();
    const setAllTags = useSetTags();
    return { allTags, setAllTags };
};

export default useTagContext;