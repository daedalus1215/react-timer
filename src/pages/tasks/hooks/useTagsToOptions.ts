import useFetchAllTags from "pages/tags/useFetchAllTags";
import useAllTagSelectors from "redux/selectors/useAllTagSelectors";

const useTagsToOptions = () => {
    useFetchAllTags();
    const tags = useAllTagSelectors();
    return tags.map(tag => ({
        value: tag._id,
        label: tag.name
    }));
};

export default useTagsToOptions;