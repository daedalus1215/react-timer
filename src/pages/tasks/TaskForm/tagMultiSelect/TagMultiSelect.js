import React from 'react';
import PropType from 'prop-types';
import { MultiSelectAdapter } from 'components';
import { useTagContext, useTagTransformer } from '../hooks';
import styles from './TagMultiSelect.module.css';

const TagMultiSelect = ({ tags, ...rest }) => {
  const { allTags } = useTagContext();
  const transformedTags = useTagTransformer(allTags);
  const initialTags = useTagTransformer(tags);
  return (
    <MultiSelectAdapter
      className={styles.select}
      data-test-id="tag-multi-select"
      name="tags"
      options={transformedTags}
      value={initialTags}
      {...rest}
    />
  );
};

TagMultiSelect.PropType = {
  tags: PropType.array,
};

export default TagMultiSelect;
