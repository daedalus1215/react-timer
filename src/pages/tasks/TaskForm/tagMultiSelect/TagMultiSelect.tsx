import React from 'react';
import { MultiSelectAdapter } from 'components';
import { useTagContext, useTagTransformer } from '../hooks';
import styles from './TagMultiSelect.module.css';

interface TagMultiSelectProps {
  tags: any;
  input: any;
}

const TagMultiSelect: React.FC<TagMultiSelectProps> = ({ tags, input, ...rest }) => {
  const { allTags } = useTagContext();
  const transformedTags = useTagTransformer(allTags);
  const initialTags = useTagTransformer(tags);

  return <MultiSelectAdapter 
            className={styles.select} 
            data-test-id="tag-multi-select" 
            name="tags" 
            input={input}
            options={transformedTags} 
            value={initialTags} 
            {...rest} />;
};

export default TagMultiSelect;
