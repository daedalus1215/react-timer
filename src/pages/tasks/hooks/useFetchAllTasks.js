import { useEffect } from 'react';
import { fetchApiData } from 'utils';
import { useLoadinSpinnerContext } from 'hooks';

//@TODO: Model this after useFetchAllTags
const useFetchAllTasks = setTasks => {
  const { setIsLoadin } = useLoadinSpinnerContext();
const tasks = localStorage.getItem('tasks');
const parsedTasks = tasks ? JSON.parse(tasks): null;

  return useEffect(() => {
    if(parsedTasks) {
      setTasks(parsedTasks);
    } else {
      // setIsLoadin(true);
      fetchApiData('tasks', {}, setTasks);
      // setIsLoadin(false);
    }
  }, [setTasks, setIsLoadin]);
};

export default useFetchAllTasks;
