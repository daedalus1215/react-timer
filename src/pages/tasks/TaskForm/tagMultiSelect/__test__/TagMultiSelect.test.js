import React from 'react';
import { render } from '@testing-library/react';
import TagMultiSelect from '../TagMultiSelect';
import { useTagContext, useTagTransformer } from '../../hooks';

jest.mock('../../hooks/useTagContext');
jest.mock('../../hooks/useTagTransformer');

describe('src/pages/createOrEditTask/form/__test__/TagMultiSelect.test.js', () => {
  describe('TagMultiSelect', () => {
    // Arrange
    const allTags = [{ id: 1 }, { id: 2 }];

    beforeEach(() => {
      useTagContext.mockReturnValue(allTags);
      useTagTransformer.mockReturnValue(allTags);
    });

    it('should display TagMultiSelect', async () => {
      // Arrange
      const input = {
        onChange: jest.fn(),
        value: 'value'
      }
      // Act
      const actual = render(<TagMultiSelect tags={allTags} input={input} />);
     
      // Assert
      expect(useTagContext).toBeCalledTimes(1);
      expect(useTagTransformer).toBeCalledTimes(3);
      expect(actual).toMatchSnapshot();
    });
  });
});
