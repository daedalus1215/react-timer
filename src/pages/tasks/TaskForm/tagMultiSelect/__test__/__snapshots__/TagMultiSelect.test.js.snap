// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src/pages/createOrEditTask/form/__test__/TagMultiSelect.test.js TagMultiSelect should display TagMultiSelect 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="go2139093995 select"
      >
        <div
          aria-labelledby="Select"
          aria-readonly="true"
          class="go2646822163 dropdown-container"
          tabindex="0"
        >
          <div
            class="go2642161244 dropdown-heading"
          >
            <div
              class="dropdown-heading-value"
            >
              <span>
                All items are selected.
              </span>
            </div>
            <button
              aria-label="Clear Selected"
              class="go1003880112 clear-selected-button"
              type="button"
            >
              <svg
                class="dropdown-search-clear-icon gray"
                fill="none"
                height="24"
                stroke="currentColor"
                stroke-width="2"
                width="24"
              >
                <line
                  x1="18"
                  x2="6"
                  y1="6"
                  y2="18"
                />
                <line
                  x1="6"
                  x2="18"
                  y1="6"
                  y2="18"
                />
              </svg>
            </button>
            <svg
              class="dropdown-heading-dropdown-arrow gray"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-width="2"
              width="24"
            >
              <path
                d="M6 9L12 15 18 9"
              />
            </svg>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="go2139093995 select"
    >
      <div
        aria-labelledby="Select"
        aria-readonly="true"
        class="go2646822163 dropdown-container"
        tabindex="0"
      >
        <div
          class="go2642161244 dropdown-heading"
        >
          <div
            class="dropdown-heading-value"
          >
            <span>
              All items are selected.
            </span>
          </div>
          <button
            aria-label="Clear Selected"
            class="go1003880112 clear-selected-button"
            type="button"
          >
            <svg
              class="dropdown-search-clear-icon gray"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-width="2"
              width="24"
            >
              <line
                x1="18"
                x2="6"
                y1="6"
                y2="18"
              />
              <line
                x1="6"
                x2="18"
                y1="6"
                y2="18"
              />
            </svg>
          </button>
          <svg
            class="dropdown-heading-dropdown-arrow gray"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-width="2"
            width="24"
          >
            <path
              d="M6 9L12 15 18 9"
            />
          </svg>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
